### PR TITLE
Internal seed of keras.layers.Dropout gets increased by +1 every time a tf.function gets created from the model

### LINKED ISSUE #17548: Internal seed of keras.layers.Dropout gets increased by +1 every time a tf.function gets created from the model
**System information**.
- Have I written custom code (as opposed to using a stock example script provided in Keras): no
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Linux CentOS
- TensorFlow installed from (source or binary): PYPI
- TensorFlow version (use command below): 2.11.0
- Python version: 3.7
- Bazel version (if compiling from source):
- GPU model and memory:
- Exact command to reproduce:

**Describe the problem**.

When using `keras.layers.Dropout` with a user defined seed, the seed gets increments by `+1` every time `tf.function` gets called (i.e. in `make_train_function`, ...).

See code and log below. As you can see, every time I run `tf.function`, the seed of the Keras dropout layer gets increased. BUT the seed of manually crafted Dropout using `tf.random.uniform` does not. So I assume it's a Keras problem. I tried to pinpoint the error and found the following code:

https://github.com/keras-team/keras/blob/d2a6b9e0efdd01dd1bc6cf0bb52336d0009aba6c/keras/backend.py#L2007-L2025

Here the seed gets increased "When user didn't provide any original seed". However, the check is wrong, as `self._seed` is the user defined seed. It get set in the constructor and also here: https://github.com/keras-team/keras/blob/d2a6b9e0efdd01dd1bc6cf0bb52336d0009aba6c/keras/backend.py#L1974-L1976
https://github.com/keras-team/keras/blob/d2a6b9e0efdd01dd1bc6cf0bb52336d0009aba6c/keras/backend.py#L2027-L2033

So I think the `make_legacy_seed` function needs to be fixed. I'm not sure about this comment "... it is important to generate different seed for stateful ops ...", because the underlying `tf.random.uniform` stores the state in the `PhiloxRandom` C++ class, which gets incremented throughout the training process and only gets reset when triggered by `tf.random.set_seed(...)`. I don't see why the seed needs to be manually increased here.

**Describe the current behavior**.
Every time we run `tf.function` the seed of the `keras.layers.Dropout` gets increased.

**Describe the expected behavior**.
The seeds should be stable, so we can have deterministic behavior.

**[Contributing](https://github.com/keras-team/keras/blob/master/CONTRIBUTING.md)**.

- Do you want to contribute a PR? (yes/no): yes
- If yes, please read [this page](https://github.com/keras-team/keras/blob/master/CONTRIBUTING.md) for instructions
- Briefly describe your candidate solution(if contributing):

Personally I would remove the `make_legacy_seed` method, but I'm not sure if there was a special purpose for this behavior. Alternative, we could store a `self.user_provided_seed = seed is not None` and check for this within `make_legacy_seed`.

**Standalone code to reproduce the issue**.

```python
import tensorflow as tf
import numpy as np

tf.random.set_seed(123)

rate = 0.5
inp = tf.keras.Input((5,))
x = tf.keras.layers.Dropout(rate=rate, seed=345)(inp)
y = tf.cast(tf.random.uniform(shape=tf.shape(inp), dtype=inp.dtype, seed=345) >= rate, inp.dtype) * inp * (1/rate)
model = tf.keras.Model(inputs=[inp], outputs=[x, y])

args = np.ones((1, 5,), np.float32)
counter = 0

def print_seeds():
    global counter
    print(f"Run #{counter}")
    counter += 1

    print("\tModel:")
    for l in model.layers:
        if hasattr(l, 'seed'):
            print("\t\t", l.name, l.seed)

    print("\tGraph:")
    func = tf.function(model.__call__)
    graph = func.get_concrete_function(args, training=True).graph
    for node in graph.get_operations():
        if node.op_def.name == "RandomUniform":
            print("\t\t", node.name, 'Seed:', node.get_attr('seed'), 'Seed2:', node.get_attr('seed2'))
    print()

print_seeds()
print_seeds()
print_seeds()
```

**Source code / logs**.

```
Run #0
        Model:
                 dropout 345
        Graph:
                 model/tf.random.uniform/random_uniform/RandomUniform Seed: 123 Seed2: 345
                 model/dropout/dropout/random_uniform/RandomUniform Seed: 123 Seed2: 345

Run #1
        Model:
                 dropout 345
        Graph:
                 model/tf.random.uniform/random_uniform/RandomUniform Seed: 123 Seed2: 345
                 model/dropout/dropout/random_uniform/RandomUniform Seed: 123 Seed2: 346

Run #2
        Model:
                 dropout 345
        Graph:
                 model/tf.random.uniform/random_uniform/RandomUniform Seed: 123 Seed2: 345
                 model/dropout/dropout/random_uniform/RandomUniform Seed: 123 Seed2: 347
```

### PR DESCRIPTION
**Pull request by mergian**
This fixes #17548

As described in the issue, the problem is that `RandomGenerator` stores a seed that gets incremented whenever `def call(...)` of a layer gets executed (either in eager mode or when converted to `tf.function`). This seed cannot be reset through any method provided by `tf.random.set_seed` or similar keras functionality.

The reasoning for this state within the `RandomGenerator` is to provide layers that use multiple random operations a method to only use one user provided seed and still be able to generate different random numbers. Here an example:

```python
# this code is just conceptional!
class DoubleDropout:
	def __init__(self, seed):
		self.gen = RandomGenerator(seed)

	def call(self, x):
		A = self.gen.random_uniform(x.shape, 0.0, 1.0, x.dtype) # gets assigned seed
		B = self.gen.random_uniform(x.shape, 0.0, 1.0, x.dtype) # gets assigned seed + 1
		return A * x + B * x
```

As can be seen without this trick both calls to `random_uniform` would generate identical random numbers. So it's necessary to have this kind of mechanism in place. However, as this state is stored within the `self.gen` object, it gets increased every time `def call(...)` gets executed. And as it cannot be reset, it's destroying determinism if the same model gets executed a second time, without restarting the entire application.

My solution is in principle something like this:

```python
def call(self, x):
	with self.gen as gen:
		A = gen.random_uniform(x.shape, 0.0, 1.0, x.dtype) # gets assigned seed
		B = gen.random_uniform(x.shape, 0.0, 1.0, x.dtype) # gets assigned seed + 1
	return A * x + B * x
```

Here I'm using a scope object, which contains this state. So the behavior within the `def call` method is identical, but it gets destroyed at the end of that method, which guarantees identical behavior between multiple calls to `def call`.

### CODE REVIEW

**Comment by fchollet** [comment]
Thanks for the PR.

> And as it cannot be reset, it's destroying determinism if the same model gets executed a second time, without restarting the entire application.

I don't quite understand how it is "destroying determinism", can you explain again but more slowly?

**Comment by qlzh727** [comment]
@fchollet, I think the issue is recorded in https://github.com/keras-team/keras/issues/17548, and let's move the conversion there.

**Comment by mergian** [comment]
In the example you posted in the Issue you are creating a second Dropout layer, which hides the effect I'm talking about. See this example:

```python
import tensorflow as tf
import numpy as np

tf.keras.utils.set_random_seed(123)

inp = tf.keras.Input((6))
x = tf.keras.layers.Dropout(0.5, seed=123)(inp)
model = tf.keras.Model(inp, x)

data = np.random.rand(1, 6)

tf.keras.utils.set_random_seed(123)

A0 = model(data, training=True).numpy()
B0 = model(data, training=True).numpy()

tf.keras.utils.set_random_seed(123)

A1 = model(data, training=True).numpy()
B1 = model(data, training=True).numpy()

print("A:")
print(A0)
print(A1)

print("B:")
print(B0)
print(B1)
```

Output:
```
A:
[[1.3929384 0.        0.4537029 0.        1.4389379 0.8462129]]
[[1.3929384 0.        0.        1.1026295 0.        0.8462129]]
B:
[[1.3929384 0.        0.        1.1026295 0.        0.8462129]]
[[0.        0.        0.4537029 1.1026295 0.        0.       ]]
```

I would expect, that after reseeding I get the exact same sequence from the model. But this is not the case, because I cannot reset the internal state of the Dropout with `tf.keras.utils.set_random_seed(123)`. So I cannot reproduce the same results with this model. I need to clone the model or even restart the application.

You can also take my script and change the seed to None. Then you will get:
```
A:
[[0.        0.5722787 0.        1.1026295 1.4389379 0.8462129]]
[[0.        0.5722787 0.        1.1026295 1.4389379 0.8462129]]
B:
[[1.3929384 0.        0.4537029 1.1026295 1.4389379 0.       ]]
[[1.3929384 0.        0.4537029 1.1026295 1.4389379 0.       ]]
```

Which in my opinion is the intended behavior. The reason for this is, that this legacy mode only gets used when the seed value is explicitly set by the user. And my fix produces the identical behavior as of `seed=None` also for seeds with integer values.

**Review by danik292 — APPROVED** [review]
The code is well programeed. I think the random is now really random😁.
